theano/gpuarray/c_code/dnn_fwd.c

```diff
@@ -167,31 +204,17 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
             cuda_exit(c->ctx);
             return 1;
           }
+          prev_algo.algo = algo;
+          // no tensor_op returned from Get()
+          prev_algo.mathType = mathtype = CUDNN_DEFAULT_MATH;
+	  // fprintf(stderr, "(cudnnGetConvolutionForwardAlgorithm: (err:%d), algo: %d\n", err, algo);
       }
-      prev_algo = algo;
-    } else {
-      algo = prev_algo;
+    } else { 
+      algo = (cudnnConvolutionFwdAlgo_t)prev_algo.algo;
+      worksize = prev_algo.wsSize;
+      mathtype = prev_algo.mathType;
     }
-
-    #ifdef DEBUG
-    if (0 != theano_enum_to_string_cudnnConvolutionFwdAlgo_t(algo, algorithm_name))
-        return 1;
-    // NB: This is printed only when algorithm is chosen at runtime.
-    if (reuse_algo)
-        fprintf(stderr, "(reused %s)\n", algorithm_name);
-    else
-        fprintf(stderr, "(using %s)\n", algorithm_name);
-    #endif
-
-    if (params->choose_once) {
-      reuse_algo = 1;
-    } else {
-      for (unsigned int i = 0; i < PyGpuArray_NDIM(input); i++) {
-        prev_img_dims[i] = PyGpuArray_DIM(input, i);
-        prev_kern_dims[i] = PyGpuArray_DIM(kerns, i);
-      }
-    }
-  }
+  } else { /* choose_algo */
```

Comment:
This cover 2 types of "problems":

1) Work around bug in cudnn.
1) Not implemented cases for some specific algo in cudnn.

We absolutly need to keep 1). We need workaround/bugfix in Theano for cudnn bug then what the cudnn release cycle give us. We can't wait for cudnn release.

For 2), specifiing manually which algo to use for each layer is pretty not fun for user. So user specify one algo for all layers most of the time. So if one algo don't support some cases, it would just crash if we don't fall back to another algo in that case.

Also, Theano hardcode a specific algo by default, we need to make sure it work in all cases, which mean also fall back to other algo if it don't support those cases. This is a sub case of 2).

This design choice could be changed, but we won't do that just before a release.

So for both reason, it should be reverted from this PR. for 2), it could be discussed after the release and we could change it if enough people agree with that. But I think in practice, people use time_once or the Theano default and after your PR, they could use the cache (possibly the new default after the release. Also to discuss).